File: backend/Services/ClaudeProxyService.cs
using System.Diagnostics;
using System.Net;
using Microsoft.Extensions.Options;
using OperationNexus.Api.Configuration;

namespace OperationNexus.Api.Services;

public class ClaudeProxyService : IClaudeProxyService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<ClaudeProxyService> _logger;
    private readonly string _baseUrl;
    private readonly string _haikuModel;
    private readonly string _sonnetModel;
    private const int MaxRetries = 5;
    private static readonly int[] BackoffSeconds = [2, 5, 10, 20, 40];

    public ClaudeProxyService(HttpClient httpClient, IOptions<ClaudeProxySettings> settings, ILogger<ClaudeProxyService> logger)
    {
        _httpClient = httpClient;
        _logger = logger;
        _baseUrl = settings.Value.BaseUrl.TrimEnd('/');
        _haikuModel = settings.Value.HaikuModel;
        _sonnetModel = settings.Value.SonnetModel;
        _httpClient.Timeout = TimeSpan.FromSeconds(settings.Value.TimeoutSeconds);
    }

    public async Task<string> ChatAsync(string model, string prompt, int maxTokens = 4096, double temperature = 0.1, CancellationToken ct = default)
    {
        for (int attempt = 0; attempt <= MaxRetries; attempt++)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/v1/chat/completions");
            request.Content = JsonContent.Create(new
            {
                model,
                max_tokens = maxTokens,
                temperature,
                messages = new[] { new { role = "user", content = prompt } }
            });

            var sw = Stopwatch.StartNew();
            var response = await _httpClient.SendAsync(request, ct);
            sw.Stop();

            if (response.StatusCode == HttpStatusCode.TooManyRequests && attempt < MaxRetries)
            {
                var retryAfter = response.Headers.RetryAfter?.Delta?.TotalSeconds
                    ?? BackoffSeconds[Math.Min(attempt, BackoffSeconds.Length - 1)];
                _logger.LogWarning("[ClaudeProxy] {Model} 429 rate-limited, waiting {Delay:F0}s (retry {Attempt}/{Max})",
                    model, retryAfter, attempt + 1, MaxRetries);
                await Task.Delay(TimeSpan.FromSeconds(retryAfter), ct);
                continue;
            }

            _logger.LogInformation("[ClaudeProxy] {Model} responded in {Ms}ms (attempt {Attempt})",
                model, sw.ElapsedMilliseconds, attempt + 1);

            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<ChatCompletionResponse>(ct);
            return result?.Choices?.FirstOrDefault()?.Message?.Content
                ?? throw new InvalidOperationException("Empty chat completion response");
        }

        throw new HttpRequestException("Claude proxy rate limit exceeded after all retries");
    }
}

file record ChatMessage(string Content);
file record ChatChoice(ChatMessage Message);
file record ChatCompletionResponse(List<ChatChoice> Choices);

File: src/main/services/claudeProxyService.ts
import { getConfig } from '../config'
import { createLogger } from './logger'

const log = createLogger('ClaudeProxy')

interface ChatMessage {
  role: string
  content: string
}

interface ChatChoice {
  message: { content: string }
}

interface ChatCompletionResponse {
  choices: ChatChoice[]
}

const MAX_RETRIES = 5
const BACKOFF_SECONDS = [2, 5, 10, 20, 40]

export const claudeProxyService = {
  async chatAsync(
    model: string,
    prompt: string,
    maxTokens = 4096,
    temperature = 0.1,
    systemPrompt?: string,
    signal?: AbortSignal
  ): Promise<string> {
    const { claudeProxy } = getConfig()
    const baseUrl = claudeProxy.baseUrl.replace(/\/+$/, '')

    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
      const messages: ChatMessage[] = []
      if (systemPrompt) messages.push({ role: 'system', content: systemPrompt })
      messages.push({ role: 'user', content: prompt })

      const response = await fetch(`${baseUrl}/v1/chat/completions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model, max_tokens: maxTokens, temperature, messages }),
        signal,
      })

      if (response.status === 429 && attempt < MAX_RETRIES) {
        const retryAfter = parseInt(response.headers.get('retry-after') ?? '', 10)
        const delay = isNaN(retryAfter) ? BACKOFF_SECONDS[Math.min(attempt, BACKOFF_SECONDS.length - 1)] : retryAfter
        log.warn(`${model} 429 rate-limited, waiting ${delay}s (retry ${attempt + 1}/${MAX_RETRIES})`)
        await new Promise(resolve => setTimeout(resolve, delay * 1000))
        continue
      }

      if (!response.ok) {
        throw new Error(`Claude proxy error ${response.status}: ${response.statusText}`)
      }

      const result = (await response.json()) as ChatCompletionResponse
      const content = result.choices?.[0]?.message?.content
      if (!content) throw new Error('Empty chat completion response')
      return content
    }

    throw new Error('Claude proxy rate limit exceeded after all retries')
  },

  async checkAvailability(): Promise<boolean> {
    const { claudeProxy } = getConfig()
    const baseUrl = claudeProxy.baseUrl.replace(/\/+$/, '')
    try {
      const response = await fetch(`${baseUrl}/v1/models`, {
        signal: AbortSignal.timeout(5000),
      })
      return response.ok
    } catch (err) {
      log.error('Availability check failed', err instanceof Error ? err : new Error(String(err)))
      return false
    }
  },
}
